lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -2390,6 +2390,43 @@ bool TypeChecker::isExplicitlyConvertibleTo(Type type1, Type type2,
                                 ConstraintKind::ExplicitConversion, dc);
 }
 
+bool TypeChecker::isDeclConvertibleTo(const ValueDecl *decl, Type fromType,
+                                      Type toType) {
+  auto *outerDC = decl->getDeclContext();
+  auto *innerDC = decl->getInnermostDeclContext();
+
+  ConstraintSystem cs(*this, outerDC, ConstraintSystemOptions());
+
+  auto openType = [innerDC, outerDC, &cs](Type type) -> Type {
+    if (type->is<GenericFunctionType>()) {
+      llvm::DenseMap<CanType, TypeVariableType *> replacements;
+      return cs.openFunctionType(type->castTo<GenericFunctionType>(),
+                                 /*numArgLabelsToRemove=*/0,
+                                 /*locator=*/nullptr, replacements, innerDC,
+                                 outerDC, /*skipProtocolSelfConstraint=*/false);
+    } else if (type->hasTypeParameter()) {
+      return cs.openType(type, /*locator=*/nullptr);
```

Comment:
What would be the condition?  If the Decl is not `GenericFunctionType`, but has a `GenericSignature`, then call `openGeneric`?  Do we still call `openType` otherwise?  `openFunctionType` calls `openType` for non-`GenericFunctionType`s.